### PR TITLE
Add setting to try Google Maps instead of Apple Maps

### PR DESCRIPTION
**Pull request by zadr**
Simulator Screen Shot - iPhone X - 2019-04-10 at 22 00 18

no idea why bitrise CI job is failing to link Realm in ¯\\_(ツ)_/¯

### CODE REVIEW

File: SF iOS/SF iOS/Helpers/Settings/NSUserDefaults+Settings.h [hunk 1]
```diff
@@ -0,0 +1,20 @@
+//
+//  NSUserDefaults+Settings.h
+//  SF iOS
+//
+//  Created by Zachary Drayer on 4/8/19.
+//  Copyright © 2019 Amit Jain. All rights reserved.
```
Comment (rodericj) [inline SF iOS/SF iOS/Helpers/Settings/NSUserDefaults+Settings.h]: Is the project settings somehow adding his name to the file head?

File: SF iOS/SF iOS/AppDelegate.m [hunk 2]
```diff
@@ -7,57 +7,34 @@
...
-    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
-    // Use this method to pause ongoing tasks, disable timers, and invalidate graphics rendering callbacks. Games should use this method to pause the game.
-}
-
-
-- (void)applicationDidEnterBackground:(UIApplication *)application {
-    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
-    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
+- (BOOL)application:(UIApplication *)application willFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
+    self.navigationContainer = [[SwipableNavigationContainer alloc] init];
```
Comment (rodericj) [inline SF iOS/SF iOS/AppDelegate.m:22]: Oh you made it a nav! Now I gotta build and run

File: SF iOS/SF iOS/EventDetails/EventDetailsViewController.m [hunk 3]
```diff
@@ -83,10 +83,15 @@ - (void)viewDidLoad {
...
-                                    usingTransportType:transportType];
-    }];
+                                [DirectionsRequest requestDirectionsToLocation:self.event.location.location
+                                                                      withName:self.event.venueName
+                                                            usingTransportType:transportType
+                                                                    completion:^(BOOL success) {
+                                                                        if (!success) {
+                                                                            // show error
+                                                                        }
+                                                                    }];
```
Comment (rodericj) [inline SF iOS/SF iOS/EventDetails/EventDetailsViewController.m:93]: We briefly discussed this but why is this necessary?

File: SF iOS/SF iOS/Settings/SettingsViewController.m [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+typedef NS_ENUM(NSInteger, SettingsSection) {
+    SettingsSectionDirectionsProvider,
+    SettingsSectionCount
+};
+
+typedef NS_ENUM(NSInteger, SettingsSectionDirectionsProviderValues) {
+    SettingsSectionDirectionsProviderApple,
+    SettingsSectionDirectionsProviderGoogle,
+    SettingsSectionDirectionsProviderCount
```
Comment (rodericj) [inline SF iOS/SF iOS/Settings/SettingsViewController.m:19]: I forgot about this “last item is the count” trick. Neat.

File: SF iOS/SF iOS/Settings/SettingsViewController.m [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    [super loadView];
+
+    [self.view addSubview:self.tableView];
+
+    [self.tableView.topAnchor constraintEqualToAnchor:self.view.topAnchor].active = true;
+    [self.tableView.leftAnchor constraintEqualToAnchor:self.view.leftAnchor].active = true;
+    [self.tableView.rightAnchor constraintEqualToAnchor:self.view.rightAnchor].active = true;
+    [self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor].active = true;
+    [self.tableView.widthAnchor constraintEqualToAnchor:self.view.widthAnchor].active = true;
+}
```
Comment (rodericj) [inline SF iOS/SF iOS/Settings/SettingsViewController.m:57]: This is a little different than the rest of the app the way you use loadView. I might try changing that others to use your way. It is a bit nicer imho

**Review by rodericj — COMMENTED** [review]
Damn that was a big PR for a 1 day effort.

File: SF iOS/SF iOS/Helpers/Settings/NSUserDefaults+Settings.h (hunk 1 above)
Comment (zadr) [reply SF iOS/SF iOS/Helpers/Settings/NSUserDefaults+Settings.h]: yup. to be honest, my eyes glaze over the comment header at the top of each file and i missed this

File: SF iOS/SF iOS/EventDetails/EventDetailsViewController.m (hunk 3 above)
Comment (zadr) [reply SF iOS/SF iOS/EventDetails/EventDetailsViewController.m:93]: its mostly for cases like where someone an app that they don't have installed, e.g.: they ask to use Google Maps but never got around to installing the app / deleted it after changing the setting

File: SF iOS/SF iOS/Settings/SettingsViewController.m (hunk 5 above)
Comment (zadr) [reply SF iOS/SF iOS/Settings/SettingsViewController.m:57]: yeah, its my preferred approach to separating out init from configuration. arguably, `viewDidLoad` could also work for setting the constraints, but it seemed silly to override it just for this ¯\\_(ツ)_/¯

File: SF iOS/SF iOS/AppDelegate.m (hunk 2 above)
Comment (zadr) [reply SF iOS/SF iOS/AppDelegate.m:22]: you don't _have_ to. but you probably should, if only to verify that things actually work :D

File: SF iOS/SF iOS/EventDetails/EventDetailsViewController.m (hunk 3 above)
Comment (zadr) [reply SF iOS/SF iOS/EventDetails/EventDetailsViewController.m:93]: will present an alert in a follow-up PR, this one was getting big enough as it is…

File: SF iOS/SF iOS/Helpers/Settings/NSUserDefaults+Settings.h (hunk 1 above)
Comment (zadr) [reply SF iOS/SF iOS/Helpers/Settings/NSUserDefaults+Settings.h]: pushed a commit to fix this + remove the copyright from files we both made

**Comment by rodericj** [comment]
@zadr can you rebase from master? The CI fix is there.

**Comment by zadr** [comment]
> @zadr can you rebase from master? The CI fix is there.

all done!

File: SF iOS/SF iOS/EventDetails/EventDetailsViewController.m (hunk 3 above)
Comment (rodericj) [reply SF iOS/SF iOS/EventDetails/EventDetailsViewController.m:93]: Right, just hit the spot where I don't have google installed. Is there a way to pre-check that they have it and not even allow the swipe experience?

File: SF iOS/SF iOS/Helpers/Settings/NSUserDefaults+Settings.h (hunk 1 above)
Comment (rodericj) [reply SF iOS/SF iOS/Helpers/Settings/NSUserDefaults+Settings.h]: Explicitly thanking you for removing it from mine too. 🙇 Saved me that anxiety when I noticed it 3 months from now.

File: SF iOS/SF iOS/AppDelegate.m (hunk 2 above)
Comment (rodericj) [reply SF iOS/SF iOS/AppDelegate.m:22]: Design feedback. I like how big the text is. I'm wondering if there is a scenario where we could change it to a scrollview or more specifically a collectionview at some point. I love keeping the context of: "Oh yes, the feed is there on the right" with a little bit of the feed showing on the right. It's a much later solution. RN I'm down to put this PR into master.

File: SF iOS/SF iOS/EventDetails/EventDetailsViewController.m (hunk 3 above)
Comment (rodericj) [reply SF iOS/SF iOS/EventDetails/EventDetailsViewController.m:93]: I thought I had said this already, but I ran into the issue where I didn't have Google Maps installed. Is there a way for us to detect specific app installation state these days with iOS? I thought you could like pre-register that or something. I can dig around.

File: SF iOS/SF iOS/AppDelegate.m [hunk 6]
```diff
@@ -7,57 +7,34 @@
...
 
-
-- (void)applicationWillTerminate:(UIApplication *)application {
-    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
+- (UIWindow *)window {
+    return self.navigationContainer.window;
 }
 
-
 @end
```
Comment (rodericj) [inline SF iOS/SF iOS/AppDelegate.m:40]: I'm really happy that you even shortened up this size of this file. I know Jerry is doing some stuff in the AppDelegate as well. I'll need to make sure his stuff is compatible with this changes too. https://github.com/ThumbWorks/sf-ios/compare/master...jerrytdev:local-notifications

File: SF iOS/SF iOS/EventDetails/EventDetailsViewController.m (hunk 3 above)
Comment (rodericj) [reply SF iOS/SF iOS/EventDetails/EventDetailsViewController.m:93]: hashtag GitHub UI messed me up.

File: SF iOS/SF iOS/Settings/SettingsViewController.m (hunk 5 above)
Comment (rodericj) [reply SF iOS/SF iOS/Settings/SettingsViewController.m:57]: Also worth noting your use of _tableView in the initializer.

Comment (frederickohen) [reply SF iOS/SF iOS/Settings/SettingsViewController.m:57]: Wow @zadr this is extremely good work. 
Question for learning purposes, what's significant about the use of `_tableView` in the initializer?  @rodericj

Comment (rodericj) [reply SF iOS/SF iOS/Settings/SettingsViewController.m:57]: There is technically a difference between calling `[_tableview someMethod]` vs `[self.tableView someMethod]`. From what I remember, calling `self.` calls a method that is auto generated through the `@property ....` in the interface. The method would be something like 

```
- (UITableView *)tableView {
   return _tableview;
}
```

^Hey that's pretty helpful right? ^


That method can be overridden in your class....or more importantly, given that Objective-C is a HIGHLY dynamic language, this can be overridden by basically anyone anywhere in the app so at build time you have no guarantee that calling `self.tableView` isn't running this method:

```
- (UITableView *)tableView {
   [GlobalAnalyticsProvider takeSnapshotOfUserRightNowAndSendItToABadActor];
   return _tableView;
}
```


So it's a safety thing. Before Zach's addition here I was being slightly reckless if I'm being honest. So using _tableView becomes a policy decision.

That said, there may be other reasons, but also I may be entirely wrong. I reserve the right to be wrong.

Comment (zadr) [reply SF iOS/SF iOS/Settings/SettingsViewController.m:57]: Basically! You nailed it with the overridden bit.

To add my own explanation (because I typed it out before rereading your comment fully 😬):

This is a bit of defensive code. The idea is that during initialization, we don’t yet know what `self` is yet.

Because self can be re-assigned, and because we don’t what any getters will do— maybe there’s a subclass somewhere we don’t know about— we should avoid doing anything that can cause references to `self` to be sent out— by triggering KVO notifications, posting notifications, or any other form of calling out to other code— before our object is fully initialized.

So, to avoid triggering these kinds of behavior, it’s something of a best practice to prefer _instanceVariable access in init and dealloc, and .property access everywhere else with ObjC.
